components/decimal/src/options.rs

```diff
@@ -4,17 +4,6 @@
 
 //! Options for [`FixedDecimalFormat`](crate::FixedDecimalFormat).
 
-/// A bag of options defining how numbers will be formatted by
-/// [`FixedDecimalFormat`](crate::FixedDecimalFormat).
-#[derive(Debug, Eq, PartialEq, Clone, Default)]
-#[non_exhaustive]
-pub struct FixedDecimalFormatOptions {
```

Comment:
Discussion: I think we should keep the `FixedDecimalFormatOptions` because it is `non_exhaustive` and we can add more things to it in the future. What those things would be, I'm not sure; I could imagine an option to enable fractional grouping, perhaps.

https://unicode-org.atlassian.net/browse/ICU-10794

Reply:
Fractional grouping might fit under `GroupingStrategy` though. In principle I am not against leaving the door open for more options, but we discussed that with @Manishearth we were having quite a bit of trouble finding things that could fit in this stage and that aren’t something about grouping.
Padding with figure spaces maybe?

Reply:
So over FFI we can't avoid having an exhaustive set of options, in which case it makes sense to collapse the struct. But *in Rust* we are still able to have an enum. If we add new options to any formatter, we can't avoid adding a new constructor on the FFI side.

Reply:
Sure, but I’m not really talking about FFI here. We should keep `FixedDecimalFormatOptions` in Rust if we think we might add other options (beyond just more values of `GroupingStrategy`); conversely if we think that we would never add such options, we should probably drop the options struct.
Hence trying to figure out whether there are conceivable non-grouping options that might apply at this stage.

Reply:
`GroupingStrategy` is an enum with 4 values

```
    Auto,
    Never,
    Always,
    Min2,
```

`FractionalGrouping` would be an enum probably with 3 values to start

```
    Auto,
    Always,
    Never,
```

These 2 enums are orthogonal, which means they make the most sense as separate options. Otherwise, we end up with a big GroupingStrategy enum that looks like

```
    Auto,
    Never,
    Always,
    Min2,
    AutoFractionAuto,
    AutoFrationNever,
    AutoFractionAlways,
    NeverFractionAuto,
    # ...
```

Reply:
I’m not sure there all of those combinations make sense, but it does sound like we could add more options, so that we should keep the Rust struct.

Onto the FFI question then. If we keep the Rust options I would tend to think we should keep the FFI options struct as well. While extending such a struct would, as you mention, be an incompatibility and would need to happen in a new major version, not all incompatible changes are equal; the actual effect on callers of such an extension would be minor, whereas adding or removing an options struct is messier.

There is that annoying diplomat issue though. Maybe we can fix it?

Reply:
I think the way we plan to do such expansions in FFI is to just add new constructors.

The effect on callers would be a bit more major when you consider binary compatibility.

Reply:
The FFI constructor could/should likely be called `new_with_grouping_strategy`

Reply:
> The effect on callers would be a bit more major when you consider binary compatibility.

Losing binary compatibility across incompatible versions seems like a given; but while lots of codebases can be rebuilt easily, the ability to do large-scale code changes while on vacation is not universal, so I’m worried about code compatibility here; we don’t want callers to have to invent some sort of `six` for ICU4X.

Of course we can have a series of deprecated FFI constructors `new_with_grouping_strategy`, `new_with_grouping_strategy_and_another_thing`, etc., as you suggest—and indeed if we want to make the other thing available through FFI between major versions we need to have a dedicated constructor.
But doing that systematically makes the FFI API quickly divergent from the Rust one (and means we lug around our history even through incompatible versions if we don’t want to break code compatibility too much).
In contrast a constructor that takes the options struct of the latest major version would keep mirroring the Rust API of that version.